reframe/core/schedulers/slurm.py

```diff
@@ -195,15 +195,34 @@ def emit_preamble(self, job):
             hint = 'multithread' if job.use_smt else 'nomultithread'
 
         for opt in job.sched_access:
-            preamble.append('%s %s' % (self._prefix, opt))
+            if not opt.strip().startswith(('-C', '--constraint')):
+                preamble.append('%s %s' % (self._prefix, opt))
+
+        constraints = []
+        option_parser = ArgumentParser()
+        option_parser.add_argument('-C', '--constraint')
+        parsed_args, _ = option_parser.parse_known_args(job.sched_access)
+        if parsed_args.constraint:
+            constraints += [parsed_args.constraint]
+
+        # NOTE: Here last of the passed --constraint job options is taken
+        # into account in order to respect the behavior of slurm.
+        parsed_args, _ = option_parser.parse_known_args(job.options)
+        if parsed_args.constraint:
+            constraints += [parsed_args.constraint]
+
+        if constraints:
+            preamble.append(
+                self._format_option(','.join(constraints), '--constraint={0}'))
 
         preamble.append(self._format_option(hint, '--hint={0}'))
         prefix_patt = re.compile(r'(#\w+)')
         for opt in job.options:
-            if not prefix_patt.match(opt):
-                preamble.append('%s %s' % (self._prefix, opt))
-            else:
-                preamble.append(opt)
+            if not opt.strip().startswith(('-C', '--constraint')):
```

Comment:
```suggestion
            if opt.strip().startswith(('-C', '--constraint')):
                # Constraints are already processed
                continue
            
            # all the rest unindented here
```

Reply:
Also are you sure that passing a tuple to `startswith()` works as you intend here?? It's not documented.

Reply:
Yes, I did not know either, it is mentioned here: https://docs.python.org/3/library/stdtypes.html#str.startswith that the prefix can be a tuple of prefixes to try.